sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala

```diff
@@ -57,12 +58,31 @@ class StatePartitionReader(
       partition.sourceOptions.operatorId, partition.partition, partition.sourceOptions.storeName)
     val stateStoreProviderId = StateStoreProviderId(stateStoreId, partition.queryId)
 
+    val allStateStoreMetadata = new StateMetadataPartitionReader(
+      partition.sourceOptions.stateCheckpointLocation.getParent.toString, hadoopConf)
+      .stateMetadata.toArray
+
+    val stateStoreMetadata = allStateStoreMetadata.filter(entry =>
+      entry.operatorId == partition.sourceOptions.operatorId
+        && entry.stateStoreName == partition.sourceOptions.storeName
+    )
+    val numColsPrefixKey = if (stateStoreMetadata.isEmpty) {
+      logWarning("Metadata for state store not found, possible cause is this checkpoint " +
+        "is created by older version of spark. The state of session window aggregation can't be " +
+        "read correctly without state metadata and runtime exception will be thrown. " +
+        "Run the streaming query in newer spark version to generate state metadata.")
+      0
+    } else {
+      require(stateStoreMetadata.length == 1)
+      stateStoreMetadata.head.numColsPrefixKey
+    }
+
     // TODO: This does not handle the case of session window aggregation; we don't have an
```

Comment:
Let's remove this TODO comment as you're addressing this.

Reply:
Removed.